File: offer_11_min_array.py
from typing import List
class Solution:
    def minArray1(self, arr: List[int]) -> int:
        for i in range(1, len(arr)):
            if arr[i] < arr[i-1]:
                return arr[i]
        return arr[0]
    
    def minArray(self, arr: List[int]) -> int:
        low, high = 0, (len(arr) - 1)
        pivot = low + (high - low)//2
        while low < high:
            if arr[pivot] < arr[high]:
                high = pivot
            elif arr[pivot] > arr[high]:
                low = pivot + 1
            else:
                high -= 1
            pivot = low + (high - low)//2
        
        return arr[low]



if __name__ == '__main__':
    s = Solution()
    print(s.minArray([4, 5, 6, 1, 2, 3]))

File: cpp/offer11_min_array.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int minArray1(vector<int> &nums)
    {
        for (int i=1; i<nums.size(); i++)
        {
            if (nums[i]<nums[i-1])
                return nums[i];
        }
        return nums[0];
    }
    int minArray(vector<int> &numbers)
    {
        int i = 0;
        int j = numbers.size() - 1;

        while (i<j)
        {
            int m = i + (j-i)/2;
            if (numbers[m] > numbers[j])
                i = m + 1;
            else if (numbers[m] < numbers[j])
                j = m;
            else
            {
                int x = i;
                for (int k=i+1; k<=j; k++)
                {
                    if (numbers[k]<numbers[x])
                        x = k;
                }
                return numbers[x];
            }
        }
        return numbers[i];
    }
};

int main()
{
    return 0;
}